providers/aws-ec2/src/test/java/org/jclouds/aws/ec2/compute/AWSEC2ComputeServiceApiMockTest.java

```diff
@@ -234,4 +243,99 @@ public void listNodesWhereImageDoesntExist() throws Exception {
       assertPosted(DEFAULT_REGION, "Action=DescribeImages&ImageId.1=ami-aecd60c7");
       assertPosted(DEFAULT_REGION, "Action=DescribeSpotInstanceRequests");
    }
+   
+   public void deleteIncidentalResourcesSuccessfully() throws Exception {
+      enqueueRegions(DEFAULT_REGION);
+      enqueueXml(DEFAULT_REGION, "/describe_securitygroups_extension_single.xml");
+      enqueueXml(DEFAULT_REGION, "/delete_securitygroup.xml");
+      enqueueXml(DEFAULT_REGION, "/describe_keypairs_jcloudssingle.xml");
+      enqueueXml(DEFAULT_REGION, "/describe_instances_empty.xml");
+      enqueueXml(DEFAULT_REGION, "/delete_keypair.xml");
+      enqueueXml(DEFAULT_REGION, "/describe_placement_groups.xml");
+      enqueueXml(DEFAULT_REGION, "/delete_placementgroup.xml");
+      enqueueXml(DEFAULT_REGION, "/describe_placement_groups_empty.xml");
+
+      AWSEC2ComputeService computeService = (AWSEC2ComputeService) computeService();
+
+      computeService.cleanUpIncidentalResources(DEFAULT_REGION, "sg-3c6ef654");
+
+      assertPosted(DEFAULT_REGION, "Action=DescribeRegions");
+      assertPosted(DEFAULT_REGION, "Action=DescribeSecurityGroups&GroupName.1=jclouds%23sg-3c6ef654");
+      assertPosted(DEFAULT_REGION, "Action=DeleteSecurityGroup&GroupName=jclouds%23sg-3c6ef654");
+      assertPosted(DEFAULT_REGION, "Action=DescribeKeyPairs&Filter.1.Name=key-name&Filter.1.Value.1=jclouds%23sg-3c6ef654%23us-east-1%2A");
+      assertPosted(DEFAULT_REGION, "Action=DescribeInstances&Filter.1.Name=instance-state-name&Filter.1.Value.1=terminated&Filter.1.Value.2=shutting-down&Filter.2.Name=key-name&Filter.2.Value.1=jclouds%23sg-3c6ef654");
+      assertPosted(DEFAULT_REGION, "Action=DeleteKeyPair&KeyName=jclouds%23sg-3c6ef654");
+      assertPosted(DEFAULT_REGION, "Action=DescribePlacementGroups&GroupName.1=jclouds%23sg-3c6ef654%23us-east-1");
+      assertPosted(DEFAULT_REGION, "Action=DeletePlacementGroup&GroupName=jclouds%23sg-3c6ef654%23us-east-1");
+      assertPosted(DEFAULT_REGION, "Action=DescribePlacementGroups&GroupName.1=jclouds%23sg-3c6ef654%23us-east-1");
+   }
+
+   public void deleteIncidentalResourcesGivingDependencyViolationForSecurityGroup() throws Exception {
+      runDeleteIncidentalResourcesGivingErrForSecurityGroup("DependencyViolation");
+   }
+   
+   public void deleteIncidentalResourcesGivingInUseForSecurityGroup() throws Exception {
+      runDeleteIncidentalResourcesGivingErrForSecurityGroup("InvalidGroup.InUse");
+   }
+   
+   protected void runDeleteIncidentalResourcesGivingErrForSecurityGroup(String errCode) throws Exception {
+      // Complicated dispatcher is needed because cleanUpIncidentalResources will retry an unpredictable 
+      // number of times (because it is time-based, for 3 seconds - not configurable).
```

Comment:
I understand what you mean, but given that you've had to struggle with this test to make it work with the current implementation (because it is not a good one), I think it is reasonable to propose to fix and improve the code as soon as we find ourselves having to workaround it.

I know this might not be directly related to the change in the error parser, but if we just *fix stuff* without trying to make the related code better, we'll just be growing the codebase in a way that will make it harder to maintain and evolve in the future. In this case, there is just a test that calls code that is not test friendly. Instead of making the test complex, it is better to fix that code, as it should be pretty straightforward.